pkg/certificate/manager.go

```diff
@@ -0,0 +1,146 @@
+package certificate
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+
+	"github.com/openservicemesh/osm/pkg/announcements"
+	"github.com/openservicemesh/osm/pkg/k8s/events"
+	"github.com/openservicemesh/osm/pkg/messaging"
+)
+
+var errNoIssuingCA = errors.New("no issuing CA")
+var errCertNotFound = errors.New("certificate not found")
+
+// NewManager creates a new CertManager with the passed CA and CA Private Key
+func NewManager(
+	ca *Certificate,
+	client client,
+	serviceCertValidityDuration time.Duration,
+	msgBroker *messaging.Broker) (*manager, error) {
+	if ca == nil {
+		return nil, errNoIssuingCA
+	}
+
+	m := &manager{
+		// The root certificate signing all newly issued certificates
+		ca:                          ca,
+		client:                      client,
+		serviceCertValidityDuration: serviceCertValidityDuration,
+		msgBroker:                   msgBroker,
+	}
+
+	// TODO(steeling) start the cert rotation here.
+
+	return m, nil
+}
+
+func (m *manager) deleteFromCache(cn CommonName) {
+	m.cache.Delete(cn)
+}
+
+func (m *manager) getFromCache(cn CommonName) *Certificate {
+	if certInterface, exists := m.cache.Load(cn); exists {
+		cert := certInterface.(*Certificate)
+		log.Trace().Msgf("Certificate found in cache SerialNumber=%s", cert.GetSerialNumber())
+		if cert.ShouldRotate() {
+			log.Trace().Msgf("Certificate found in cache but has expired SerialNumber=%s", cert.GetSerialNumber())
+			return nil
+		}
```

Comment:
`getFromCache` is called in `IssueCertificate`. I thought `ShouldRotate()` was used instead of the cert's expiration directly to avoid returning a certificate that would need to be rotated soon after being issued.

Reply:
I agree with Jon here, I think returning a soon-to-expired cert is better than no cert.

Particularly, it allows us to bump up the ShouldRotate time to say 1/3 of total lifetime of cert, without affecting this.

Reply:
Actually I spoke to soon. It may make sense to trigger a re-issue on certificates if it is expired, even on a GetCertificate call. Will dig through this a bit more, but since this PR isn't changing existing functionality I propose we don't block. Thoughts?

Will keep this open for further discussion